File: Tyche.Android/pool/src/main/java/com/felipearpa/tyche/pool/gamblerscore/GamblerScoreListViewModel.kt
package com.felipearpa.tyche.pool.gamblerscore

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import androidx.paging.cachedIn
import com.felipearpa.data.pool.application.GetPoolGamblerScoresByPoolUseCase
import com.felipearpa.tyche.core.emptyString
import com.felipearpa.tyche.pool.PoolGamblerScorePagingSource
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.stateIn

private const val PAGE_SIZE = 50
private const val PREFETCH_DISTANCE = 5

class GamblerScoreListViewModel @AssistedInject constructor(
    @Assisted("poolId") private val poolId: String,
    @Assisted("gamblerId") val gamblerId: String,
    private val getPoolGamblerScoresByPoolUseCase: GetPoolGamblerScoresByPoolUseCase
) : ViewModel() {

    val pageSize = PAGE_SIZE
    private var _searchText = MutableStateFlow(emptyString())

    @OptIn(ExperimentalCoroutinesApi::class)
    val poolGamblerScores = _searchText.flatMapLatest { newSearchText ->
        buildPager(searchText = newSearchText).flow.cachedIn(scope = viewModelScope)
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.Lazily,
        initialValue = PagingData.empty()
    )

    private fun buildPager(searchText: String) =
        Pager(
            config = PagingConfig(
                pageSize = pageSize,
                prefetchDistance = PREFETCH_DISTANCE,
                enablePlaceholders = true
            ),
            pagingSourceFactory = {
                PoolGamblerScorePagingSource(
                    pagingQuery = { next ->
                        getPoolGamblerScoresByPoolPagingQuery(
                            next = next,
                            poolId = poolId,
                            search = { searchText.ifEmpty { null } },
                            getPoolGamblerScoresByPoolUseCase = getPoolGamblerScoresByPoolUseCase
                        )
                    }
                )
            }
        )

    fun search(searchText: String) {
        _searchText.value = searchText.trim()
    }
}

File: iOS/Pool/Sources/Pool/GamblerScore/GamblerScoreListViewModel.swift
import Foundation
import Core
import UI
import LazyPaging
import DataPool

public class GamblerScoreListViewModel: ObservableObject {
    private let getPoolGamblerScoresByPoolUseCase : GetPoolGamblerScoresByPoolUseCase
    let gamblerId: String
    private let poolId: String
    private var searchText: String? = nil
    private var pagingSource: LazyPagingCursorSource<PoolGamblerScoreModel>!

    @MainActor
    lazy var lazyPager: LazyPaging.LazyPagingItems<String, PoolGamblerScoreModel> = {
        LazyPaging.LazyPagingItems(
            pager: Pager(
                config: LazyPaging.PagingConfig(pageSize: 50, prefetchDistance: 5, enablePlaceholders: false),
                pagingSourceFactory: { [pagingSource = self.pagingSource!] in pagingSource }
            )
        )
    }()

    public init(
        getPoolGamblerScoresByPoolUseCase : GetPoolGamblerScoresByPoolUseCase,
        gamblerId: String,
        poolId: String
    ) {
        self.getPoolGamblerScoresByPoolUseCase = getPoolGamblerScoresByPoolUseCase
        self.gamblerId = gamblerId
        self.poolId = poolId

        self.pagingSource = LazyPagingCursorSource(
            pagingQuery: { [unowned self] next in
                await getPoolGamblerScoresByPoolUseCase.execute(
                    poolId: self.poolId,
                    next: next,
                    searchText: self.searchText
                )
                .map { page in page.map { poolGamblerScore in poolGamblerScore.toPoolGamblerScoreModel() }}
                .mapError { error in error.toLocalizedError() }
            }
        )
    }

    func search(_ newSearchText: String) {
        searchText = newSearchText
        pagingSource.invalidate()
    }
}

private extension Error {
    func toLocalizedError() -> Error {
        if let networkError = self as? NetworkError {
            return networkError.toNetworkLocalizedError()
        }

        return UnknownLocalizedError()
    }
}
